sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,315 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class ProactiveOpenConnectionsProcessor implements Closeable {
+
+    private static final Logger logger = LoggerFactory.getLogger(ProactiveOpenConnectionsProcessor.class);
+    private Sinks.Many<OpenConnectionTask> openConnectionsTaskSink;
+    private Sinks.Many<OpenConnectionTask> openConnectionsTaskSinkBackUp;
+    private final ConcurrentHashMap<String, List<OpenConnectionTask>> endpointsUnderMonitorMap;
+    private final AtomicReference<WarmupFlowAggressivenessHint> aggressivenessHint;
+    private final AtomicReference<Boolean> isClosed = new AtomicReference<>(false);
+    private final AtomicBoolean isConcurrencySwitchFlow = new AtomicBoolean(false);
+    private static final Map<WarmupFlowAggressivenessHint, ConcurrencyConfiguration> concurrencySettings = new HashMap<>();
+    private final IOpenConnectionsHandler openConnectionsHandler;
+    private final RntbdEndpoint.Provider endpointProvider;
+    private Disposable openConnectionBackgroundTask;
+    private final Duration aggressiveConnectionEstablishmentDuration;
+    private final Sinks.EmitFailureHandler serializedEmitFailureHandler;
+    private static final int OPEN_CONNECTION_SINK_BUFFER_SIZE = 100_000;
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider) {
+        this(endpointProvider, null);
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider, Duration aggressiveConnectionEstablishmentDuration) {
+        this.openConnectionsTaskSink = Sinks.many().multicast().onBackpressureBuffer(OPEN_CONNECTION_SINK_BUFFER_SIZE);
+        this.openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer(OPEN_CONNECTION_SINK_BUFFER_SIZE);
+        this.aggressivenessHint = new AtomicReference<>(WarmupFlowAggressivenessHint.AGGRESSIVE);
+        this.endpointsUnderMonitorMap = new ConcurrentHashMap<>();
+        this.openConnectionsHandler = new RntbdOpenConnectionsHandler(endpointProvider);
+        this.endpointProvider = endpointProvider;
+        this.aggressiveConnectionEstablishmentDuration = aggressiveConnectionEstablishmentDuration;
+        this.serializedEmitFailureHandler = new SerializedEmitFailureHandler();
+        concurrencySettings.put(WarmupFlowAggressivenessHint.AGGRESSIVE, new ConcurrencyConfiguration(Configs.getAggressiveWarmupConcurrency(), Configs.getAggressiveWarmupConcurrency()));
+        concurrencySettings.put(WarmupFlowAggressivenessHint.DEFENSIVE, new ConcurrencyConfiguration(Configs.getDefensiveWarmupConcurrency(), Configs.getDefensiveWarmupConcurrency()));
+    }
+
+    public void init() {
+        Flux<Integer> backgroundOpenConnectionsSinkReInstantiationTask = Flux
+                .just(1)
+                .delayElements(aggressiveConnectionEstablishmentDuration);
```

Comment:
wrapSourceFluxAndSoftCompleteAfterTimeout can be null now? earlier the default is 0, now is null, it might fail here